### PR TITLE
Enable the reporting of cross-entropy or nll loss value when training CNN network using the models defined by example/image-classification

### PR DESCRIPTION
**Pull request by juliusshufan**
## Description ##
MxNET already implements the loss computation at Python layer (in python/mxnet/metric.py) 
For most of the CNN models used for image-classification, as softmax is commonly used as the output, the cross-entropy or likelihood loss value is helpful to monitor the convergence trend during training.
The current implementation of example/image-classification/fit.py already provides an extensible implementation to report some useful information during training, such as accuracy. 
My submission utilizes this design and enabling the report of cross-entropy or nll loss value during training.
 
### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
All the code changes are in example/image-classification/common/fit.py
A new argument '--loss' is introduced, and it can be set as 'ce' or 'nll-loss' corresponding to cross-entropy loss and negative likelihood loss respectively.
Taking the train_cifar10.py as an example, the loss value will be reported if the below **bold code line** added,
parser.set_defaults(
    ......
    **loss           = 'ce'**
    ......
)

The output log will be the following pattern:

INFO:root:Epoch[0] Batch [380]  Speed: 504.42 samples/sec       accuracy=0.087109       top_k_accuracy_5=0.469531       **cross-entropy=2.305971**
INFO:root:Epoch[0] Train-accuracy=0.098437
INFO:root:Epoch[0] Train-top_k_accuracy_5=0.506250
INFO:root:Epoch[0] **Train-cross-entropy=2.304529**
INFO:root:Epoch[0] Time cost=100.989
INFO:root:Epoch[0] Validation-accuracy=0.098101
INFO:root:Epoch[0] Validation-top_k_accuracy_5=0.486946
INFO:root:Epoch[0] **Validation-cross-entropy=2.302873**

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by piiswrong** [comment]
This seems too ad-hoc. Its better to allow the user to select which losses they want to see one by one with a single option. Like `--loss=acc,nll,ce...`

File: example/image-classification/common/fit.py [hunk 1]
```diff
@@ -260,6 +262,18 @@ def fit(args, network, data_loader, **kwargs):
         eval_metrics.append(mx.metric.create(
             'top_k_accuracy', top_k=args.top_k))
 
+    supported_loss = ['ce', 'nll_loss']
+    if args.loss:
```
Comment (cjolivier01) [inline example/image-classification/common/fit.py]: would len(args.loss) > 0 or come other non-implied boolean test read more easily?

File: example/image-classification/common/fit.py [hunk 2]
```diff
@@ -117,6 +117,8 @@ def add_fit_args(parser):
                        help='load the model on an epoch using the model-load-prefix')
     train.add_argument('--top-k', type=int, default=0,
                        help='report the top-k accuracy. 0 means no report.')
+    train.add_argument('--loss', type=str,
+                       help='report the cross-entropy or nll-loss. ce means cross-entropy, nll-loss means likelihood loss')
```
Comment (cjolivier01) [inline example/image-classification/common/fit.py]: This seems a little confusing to me, as in what's the actual expected loss string.
Maybe something like:
"report the loss. Options are ce (cross-entropy) or nll-loss (likelihood loss)
?

?

Comment (juliusshufan) [reply example/image-classification/common/fit.py]: @piiswrong @cjolivier01 Thanks for all your comments, and sorry for the confusion.

Let me try to explain:
1. My purpose is to report the loss value during training, this is because the loss value trends is helpful to monitor the convergence trend.
2. Per my understanding is, the cross-entropy loss and (negative) likelihood loss are most common used for softmax output, and therefore I choose reporting either cross-entropy or negative likelihood loss as long as the output layer is the softmax.

The current implementation of my PR is add a string-type argument, that is the "--loss", and the corresponding type loss will be reported accordingly.

Meanwhile, my understanding to Eric's comment is, it be better to set the the argument as a list including all the matching loss type (['ce', 'nll-loss'] for softmax), and give the user to decide which loss value will be used. 

May I know which method is acceptable? I'll then update my implementation accordingly.
Thanks for your time.

Comment (juliusshufan) [reply example/image-classification/common/fit.py]: @cjolivier01 As my repo is out-of-date to the current mxnet master, I close this PR and submit a new one https://github.com/apache/incubator-mxnet/pull/9805, in which I address your review comments and provide a more detailed description of the PR. Could you please check the PR9805 accordingly? Thanks for your time.

**Comment by juliusshufan** [comment]
As my repo is out-of-date to the current mxnet master, I close this PR and submit a new one #9805 for same purpose, in which I address your review comments and provide a more detailed description of the PR.
